### PR TITLE
Fix Issue 7006 - std.math.pow (integral, integral) crashes on negative exponents

### PR DESCRIPTION
**Pull request by berni44**
I improved the docs too, including the public unittest.

### CODE REVIEW

File: std/math.d [hunk 1]
```diff
@@ -7954,10 +7969,32 @@ if (isIntegral!(F) && isIntegral!(G))
         }
         break;
     }
+
```
Comment (thewilsonator) [inline std/math.d]: Unneeded.

File: std/math.d [hunk 2]
```diff
@@ -7912,20 +7912,35 @@ if (isFloatingPoint!(F) && isIntegral!(G))
...
+ *     division by zero.
  */
 typeof(Unqual!(F).init * Unqual!(G).init) pow(F, G)(F x, G n) @nogc @trusted pure nothrow
 if (isIntegral!(F) && isIntegral!(G))
 {
-    if (n<0) return x/0; // Only support positive powers
     typeof(return) p, v = void;
     Unqual!G m = n;
 
+    if (x == -1) return cast(typeof(return)) (m % 2 == 0 ? 1 : -1);
```
Comment (ibuclaw) [inline std/math.d]: `m & 1` is more efficient.

Comment (berni44) [reply std/math.d]: Isn't that a job for the optimizer? ;-) Anyway, I changed it.

**Comment by berni44** [comment]
I do not understand, why this always fails on ci/circleci. First I thought it is unrelated, but meanwhile the error was reproduced with each run: std/range/package.test fails. When I run that test locally it succeeds (but needs about 6 minutes). What does ci/circleci actually do?

File: std/math.d [hunk 3]
```diff
@@ -7912,20 +7912,35 @@ if (isFloatingPoint!(F) && isIntegral!(G))
...
+ * Compute the power of two integral numbers.
+ *
+ * Params:
+ *     x = base
+ *     n = exponent
+ *
+ * Returns:
+ *     x raised to the power of n. If n is negative the result is 1 / pow(x, -n),
+ *     which is calculated as integer division with remainder. This may result in
+ *     a division by zero error.
```
Comment (ahmetsait) [inline std/math.d]: I think `1 / pow(x, -n)` supposed to be `1 / pow(x, n)`

**Comment by thewilsonator** [comment]
Circle is green now, but `mir/algorithm` still fails consistently.

**Comment by berni44** [comment]
> Circle is green now, but `mir/algorithm` still fails consistently.

Found (and fixed) the problem: `m <= -1` is not a good idea, when `m` is unsigned. I added also a unittest (`assert(pow(0,uint.max) == 0);`) to guard for that problem. (There are two more expressions I put into `static if`, but that is mere for extra security; IMHO they could not break and therefore I could not add a guarding unittest for them.)

**Pull request merged**